File: Sources/MesonAnalyze/OptionDiagnosticVisitor.swift
import Logging
import MesonAST

class OptionDiagnosticVisitor: CodeVisitor {
  private static let LOG = Logger(label: "MesonAnalyze::OptionDiagnosticVisitor")
  private let metadata: MesonMetadata

  init(_ metadata: MesonMetadata) { self.metadata = metadata }

  public func visitSourceFile(file: SourceFile) { file.visitChildren(visitor: self) }

  public func visitBuildDefinition(node: BuildDefinition) { node.visitChildren(visitor: self) }

  public func visitErrorNode(node: ErrorNode) { node.visitChildren(visitor: self) }

  public func visitSelectionStatement(node: SelectionStatement) {
    node.visitChildren(visitor: self)
  }

  public func visitBreakStatement(node: BreakNode) { node.visitChildren(visitor: self) }

  public func visitContinueStatement(node: ContinueNode) { node.visitChildren(visitor: self) }

  public func visitIterationStatement(node: IterationStatement) {
    node.visitChildren(visitor: self)
  }

  public func visitAssignmentStatement(node: AssignmentStatement) {
    node.visitChildren(visitor: self)
  }

  private func checkName(_ sl: StringLiteral) {
    let contents = sl.contents()
    for c in contents {
      let alpha = c.isLetter || c.isNumber
      if !alpha && c != "_" && c != "-" {
        self.metadata.registerDiagnostic(
          node: sl,
          diag: MesonDiagnostic(
            sev: .error,
            node: sl,
            message: "Invalid chars in name: Expected `a-z`, `A-Z`, `0-9`, `-` or `_`"
          )
        )
        return
      }
    }
  }

  private func isInteger(_ dv: Node?) -> Bool {
    guard let dv = dv else { return false }
    if dv is IntegerLiteral {
      return true
    } else if let unaryExpr = dv as? UnaryExpression, let unOp = unaryExpr.op, unOp == .minus,
      unaryExpr.expression is IntegerLiteral
    {
      return true
    }
    return false
  }

  private func extractInteger(_ dv: Node?) -> Int? {
    guard let dv = dv else { return nil }
    if let dvi = dv as? IntegerLiteral {
      return dvi.parse()
    } else if let unaryExpr = dv as? UnaryExpression, let unOp = unaryExpr.op, unOp == .minus,
      let dvi = unaryExpr.expression as? IntegerLiteral
    {
      return -dvi.parse()
    }
    return nil
  }

  // Refactor this!
  // swiftlint:disable cyclomatic_complexity
  public func visitFunctionExpression(node: FunctionExpression) {
    self.metadata.registerFunctionCall(call: node)
    node.visitChildren(visitor: self)
    guard let idExpr = node.id as? IdExpression else { return }
    if idExpr.id != "option" {
      self.metadata.registerDiagnostic(
        node: node,
        diag: MesonDiagnostic(
          sev: .error,
          node: node,
          message: "Invalid function name in meson options file: \(idExpr.id)"
        )
      )
      return
    }
    guard let al = node.argumentList as? ArgumentList else {
      self.metadata.registerDiagnostic(
        node: node,
        diag: MesonDiagnostic(
          sev: .error,
          node: node,
          message: "Missing arguments in call to `option`"
        )
      )
      return
    }
    guard let nameNode = al.getPositionalArg(idx: 0) else {
      self.metadata.registerDiagnostic(
        node: node,
        diag: MesonDiagnostic(sev: .error, node: node, message: "Missing option name")
      )
      return
    }
    guard let nameNodeSl = nameNode as? StringLiteral else {
      self.metadata.registerDiagnostic(
        node: nameNode,
        diag: MesonDiagnostic(sev: .error, node: nameNode, message: "Expected string literal!")
      )
      return
    }
    Self.LOG.info("Found option: \(nameNodeSl.contents())")
    self.checkName(nameNodeSl)

    guard let optionType = al.getKwarg(name: "type") else {
      self.metadata.registerDiagnostic(
        node: nameNode,
        diag: MesonDiagnostic(sev: .error, node: nameNode, message: "Missing option type")
      )
      return
    }
    guard let optionTypeSL = optionType as? StringLiteral else {
      self.metadata.registerDiagnostic(
        node: optionType,
        diag: MesonDiagnostic(
          sev: .error,
          node: optionType,
          message: "Expected option type to be a string literal"
        )
      )
      return
    }
    let defaultValue = al.getKwarg(name: "value")
    switch optionTypeSL.contents() {
    case "string":
      if let dv = defaultValue {
        if !(dv is StringLiteral) {
          self.metadata.registerDiagnostic(
            node: dv,
            diag: MesonDiagnostic(sev: .error, node: dv, message: "Expected string literal")
          )
        }
      }
    case "integer":
      var parsed: Int?
      if let dv = defaultValue {
        if let dvi = dv as? IntegerLiteral {
          parsed = dvi.parse()
        } else if let unaryExpr = dv as? UnaryExpression, let unOp = unaryExpr.op, unOp == .minus,
          let dvi = unaryExpr.expression as? IntegerLiteral
        {
          parsed = -dvi.parse()
        } else {
          self.metadata.registerDiagnostic(
            node: dv,
            diag: MesonDiagnostic(sev: .error, node: dv, message: "Expected integer literal")
          )
        }
      }
      let minNode = al.getKwarg(name: "min")
      let maxNode = al.getKwarg(name: "max")
      if minNode != nil {
        if !isInteger(minNode) {
          self.metadata.registerDiagnostic(
            node: minNode!,
            diag: MesonDiagnostic(sev: .error, node: minNode!, message: "Expected integer literal")
          )
        }
      }
      if maxNode != nil {
        if !isInteger(maxNode) {
          self.metadata.registerDiagnostic(
            node: maxNode!,
            diag: MesonDiagnostic(sev: .error, node: maxNode!, message: "Expected integer literal")
          )
        }
      }
      if let minV = self.extractInteger(minNode), let maxV = self.extractInteger(maxNode) {
        if parsed != nil {
          if parsed! < minV {
            self.metadata.registerDiagnostic(
              node: defaultValue!,
              diag: MesonDiagnostic(
                sev: .warning,
                node: defaultValue!,
                message: "Default value (\(parsed!)) is less than the minimum value (\(minV))"
              )
            )
          } else if parsed! > maxV {
            self.metadata.registerDiagnostic(
              node: defaultValue!,
              diag: MesonDiagnostic(
                sev: .warning,
                node: defaultValue!,
                message: "Default value (\(parsed!)) is more than the maximum value (\(maxV))"
              )
            )
          }
        }
        if maxV < minV {
          self.metadata.registerDiagnostic(
            node: maxNode!,
            diag: MesonDiagnostic(
              sev: .warning,
              node: maxNode!,
              message: "Maximum value is less than the minimum value"
            )
          )
        }
      } else if let minV = self.extractInteger(minNode) {
        if parsed != nil {
          if parsed! < minV {
            self.metadata.registerDiagnostic(
              node: defaultValue!,
              diag: MesonDiagnostic(
                sev: .warning,
                node: defaultValue!,
                message: "Default value (\(parsed!)) is less than the minimum value (\(minV))"
              )
            )
          }
        }
      } else if let maxV = self.extractInteger(maxNode) {
        if parsed != nil {
          if parsed! > maxV {
            self.metadata.registerDiagnostic(
              node: defaultValue!,
              diag: MesonDiagnostic(
                sev: .warning,
                node: defaultValue!,
                message: "Default value (\(parsed!)) is more than the maximum value (\(maxV))"
              )
            )
          }
        }
      }
    case "boolean":
      if let dv = defaultValue {
        if !(dv is StringLiteral || dv is BooleanLiteral) {
          self.metadata.registerDiagnostic(
            node: dv,
            diag: MesonDiagnostic(
              sev: .error,
              node: dv,
              message: "Expected boolean or string literal"
            )
          )
        } else if dv is StringLiteral {
          let contents = (dv as! StringLiteral).contents()
          if contents == "true" || contents == "false" {
            self.metadata.registerDiagnostic(
              node: dv,
              diag: MesonDiagnostic(
                sev: .warning,
                node: dv,
                message: "String literals as value for boolean options are deprecated"
              )
            )
          } else {
            self.metadata.registerDiagnostic(
              node: dv,
              diag: MesonDiagnostic(
                sev: .warning,
                node: dv,
                message: "Boolean options must have boolean values"
              )
            )
          }
        }
      }
    case "combo":
      var parsed: String?
      if let dv = defaultValue {
        if !(dv is StringLiteral) {
          self.metadata.registerDiagnostic(
            node: dv,
            diag: MesonDiagnostic(sev: .error, node: dv, message: "Expected string literal")
          )
        } else {
          parsed = (dv as! StringLiteral).contents()
        }
      }
      var setChoices: [String] = []
      if let choices = al.getKwarg(name: "choices") {
        if let arr = choices as? ArrayLiteral {
          for val in arr.args {
            if let sl = val as? StringLiteral {
              let contents = sl.contents()
              if setChoices.contains(contents) {
                self.metadata.registerDiagnostic(
                  node: val,
                  diag: MesonDiagnostic(
                    sev: .error,
                    node: val,
                    message: "Duplicate choice: '\(contents)'"
                  )
                )
              }
              setChoices.append(contents)
            } else {
              self.metadata.registerDiagnostic(
                node: val,
                diag: MesonDiagnostic(sev: .error, node: val, message: "Expected string")
              )
            }
          }
          if setChoices.isEmpty {
            self.metadata.registerDiagnostic(
              node: arr,
              diag: MesonDiagnostic(sev: .error, node: arr, message: "Expected at least one choice")
            )
          }
        } else {
          self.metadata.registerDiagnostic(
            node: choices,
            diag: MesonDiagnostic(sev: .error, node: choices, message: "Expected a list of strings")
          )
        }
      } else {
        self.metadata.registerDiagnostic(
          node: node,
          diag: MesonDiagnostic(
            sev: .error,
            node: node,
            message: "Missing keyword 'choices' for combo-option"
          )
        )
      }
      if let s = parsed {
        if !setChoices.contains(s) {
          self.metadata.registerDiagnostic(
            node: defaultValue!,
            diag: MesonDiagnostic(
              sev: .warning,
              node: defaultValue!,
              message: "Default value ('\(s)') is not in the set of choices: \(setChoices)"
            )
          )
        }
      }
    case "array":
      var values: [String] = []
      if defaultValue != nil {
        if let arr = defaultValue as? ArrayLiteral {
          for val in arr.args {
            if let sl = val as? StringLiteral {
              let contents = sl.contents()
              if values.contains(contents) {
                self.metadata.registerDiagnostic(
                  node: val,
                  diag: MesonDiagnostic(
                    sev: .error,
                    node: val,
                    message: "Duplicate value: '\(contents)'"
                  )
                )
              }
              values.append(contents)
            } else {
              self.metadata.registerDiagnostic(
                node: val,
                diag: MesonDiagnostic(sev: .error, node: val, message: "Expected string")
              )
            }
          }
        } else {
          self.metadata.registerDiagnostic(
            node: defaultValue!,
            diag: MesonDiagnostic(
              sev: .error,
              node: defaultValue!,
              message: "Expected a list of strings"
            )
          )
        }
      }
      var choicesArr: [String] = []
      if let choices = al.getKwarg(name: "choices") {
        if let arr = choices as? ArrayLiteral {
          for val in arr.args {
            if let sl = val as? StringLiteral {
              let contents = sl.contents()
              if choicesArr.contains(contents) {
                self.metadata.registerDiagnostic(
                  node: val,
                  diag: MesonDiagnostic(
                    sev: .error,
                    node: val,
                    message: "Duplicate choice: '\(contents)'"
                  )
                )
              }
              choicesArr.append(contents)
            } else {
              self.metadata.registerDiagnostic(
                node: val,
                diag: MesonDiagnostic(sev: .error, node: val, message: "Expected string")
              )
            }
          }
        } else {
          self.metadata.registerDiagnostic(
            node: choices,
            diag: MesonDiagnostic(sev: .error, node: choices, message: "Expected a list of strings")
          )
        }
      }
    // We should check the choicesArr against the other valueArr.
    case "feature":
      if let dv = defaultValue {
        if let sl = defaultValue as? StringLiteral {
          let contents = sl.contents()
          if contents != "enabled" && contents != "disabled" && contents != "auto" {
            self.metadata.registerDiagnostic(
              node: dv,
              diag: MesonDiagnostic(
                sev: .error,
                node: dv,
                message: "Expected one of: 'enabled', 'disabled', 'auto'"
              )
            )
          }
        }
      }
    default:
      self.metadata.registerDiagnostic(
        node: optionType,
        diag: MesonDiagnostic(
          sev: .error,
          node: optionType,
          message: "Unknown option type: '\(optionTypeSL.contents())'"
        )
      )

    }
  }
  // swiftlint:enable cyclomatic_complexity

  public func visitArgumentList(node: ArgumentList) { node.visitChildren(visitor: self) }

  public func visitKeywordItem(node: KeywordItem) { node.visitChildren(visitor: self) }

  public func visitConditionalExpression(node: ConditionalExpression) {
    node.visitChildren(visitor: self)
  }

  public func visitUnaryExpression(node: UnaryExpression) { node.visitChildren(visitor: self) }

  public func visitSubscriptExpression(node: SubscriptExpression) {
    node.visitChildren(visitor: self)
  }

  public func visitMethodExpression(node: MethodExpression) { node.visitChildren(visitor: self) }

  public func visitIdExpression(node: IdExpression) { node.visitChildren(visitor: self) }

  public func visitBinaryExpression(node: BinaryExpression) { node.visitChildren(visitor: self) }

  public func visitStringLiteral(node: StringLiteral) {
    self.metadata.registerStringLiteral(node: node)
    node.visitChildren(visitor: self)
  }

  public func visitArrayLiteral(node: ArrayLiteral) { node.visitChildren(visitor: self) }

  public func visitBooleanLiteral(node: BooleanLiteral) { node.visitChildren(visitor: self) }

  public func visitIntegerLiteral(node: IntegerLiteral) { node.visitChildren(visitor: self) }

  public func visitDictionaryLiteral(node: DictionaryLiteral) { node.visitChildren(visitor: self) }

  public func visitKeyValueItem(node: KeyValueItem) { node.visitChildren(visitor: self) }
}

File: src/libanalyze/options/optiondiagnosticvisitor.cpp
#include "optiondiagnosticvisitor.hpp"

#include "log.hpp"
#include "mesonmetadata.hpp"
#include "node.hpp"

#include <cctype>
#include <cstdint>
#include <optional>
#include <set>
#include <string>

static Logger LOG("OptionDiagnosticVisitor"); // NOLINT

void OptionDiagnosticVisitor::visitArgumentList(ArgumentList *node) {
  node->visitChildren(this);
}

void OptionDiagnosticVisitor::visitArrayLiteral(ArrayLiteral *node) {
  node->visitChildren(this);
}

void OptionDiagnosticVisitor::visitAssignmentStatement(
    AssignmentStatement *node) {
  node->visitChildren(this);
}

void OptionDiagnosticVisitor::visitBinaryExpression(BinaryExpression *node) {
  node->visitChildren(this);
}

void OptionDiagnosticVisitor::visitBooleanLiteral(BooleanLiteral *node) {
  node->visitChildren(this);
}

void OptionDiagnosticVisitor::visitBuildDefinition(BuildDefinition *node) {
  node->visitChildren(this);
}

void OptionDiagnosticVisitor::visitConditionalExpression(
    ConditionalExpression *node) {
  node->visitChildren(this);
}

void OptionDiagnosticVisitor::visitDictionaryLiteral(DictionaryLiteral *node) {
  node->visitChildren(this);
}

void OptionDiagnosticVisitor::checkName(StringLiteral *sl) {
  const auto &contents = sl->id;
  if (this->options.contains(contents)) {
    this->metadata->registerDiagnostic(
        sl, Diagnostic(Severity::Error, sl, "Duplicate option: " + contents));
  }
  if (this->state.findOption(contents)) {
    this->metadata->registerDiagnostic(
        sl, Diagnostic(Severity::Error, sl,
                       "Declaration of reserved option: " + contents));
  }
  this->options.insert(contents);
  for (const auto chr : contents) {
    if ((std::isalnum(chr) != 0) || chr == '_' || chr == '-') {
      continue;
    }
    this->metadata->registerDiagnostic(
        sl,
        Diagnostic(
            Severity::Error, sl,
            "Invalid chars in name: Expected `a-z`, `A-Z`, `0-9`, `-` or `_`"));
    return;
  }
}

void OptionDiagnosticVisitor::validateStringOption(Node *defaultValue) const {
  if (dynamic_cast<StringLiteral *>(defaultValue)) {
    return;
  }
  this->metadata->registerDiagnostic(
      defaultValue,
      Diagnostic(Severity::Error, defaultValue, "Expected string literal"));
}

std::optional<int64_t>
OptionDiagnosticVisitor::parseString(const Node *node) const {
  const auto *asSL = dynamic_cast<const StringLiteral *>(node);
  if (!asSL) {
    return std::nullopt;
  }
  const auto &value = asSL->id;
  if (value.empty()) {
    return std::nullopt;
  }
  try {
    int64_t ret = 0;
    if (value.starts_with("0x") || value.starts_with("0X")) {
      ret = std::stoll(value, nullptr, 16);
    } else if (value.starts_with("0b") || value.starts_with("0B")) {
      ret = std::stoll(value.substr(2), nullptr, 2);
    } else if (value.starts_with("0O") || value.starts_with("0o")) {
      ret = std::stoll(value.substr(2), nullptr, 8);
    } else {
      ret = std::stoll(value);
    }
    this->metadata->registerDiagnostic(
        node, Diagnostic(Severity::Warning, node,
                         "String literals as value where integers are "
                         "expected, are deprecated"));
    return ret;
  } catch (...) {
    this->metadata->registerDiagnostic(
        node, Diagnostic(Severity::Error, node, "Unable to parse as integer"));
  }
  return std::nullopt;
}

std::optional<int64_t> OptionDiagnosticVisitor::parseInt(const Node *node) {
  if (const auto *integerLit = dynamic_cast<const IntegerLiteral *>(node)) {
    return integerLit->valueAsInt;
  }
  if (const auto *unaryExpr = dynamic_cast<const UnaryExpression *>(node)) {
    if (unaryExpr->op != UnaryOperator::UnaryMinus) {
      return std::nullopt;
    }
    auto parsed = this->parseInt(unaryExpr->expression.get());
    if (parsed.has_value()) {
      return -parsed.value();
    }
  }
  return std::nullopt;
}

void OptionDiagnosticVisitor::validateBooleanOption(Node *defaultValue) const {
  if (dynamic_cast<BooleanLiteral *>(defaultValue)) {
    return;
  }
  const auto *sl = dynamic_cast<StringLiteral *>(defaultValue);
  if (!sl) {
    this->metadata->registerDiagnostic(
        defaultValue, Diagnostic(Severity::Error, defaultValue,
                                 "Expected boolean value for boolean option"));
    return;
  }
  if (sl->id != "true" && sl->id != "false") {
    this->metadata->registerDiagnostic(
        defaultValue, Diagnostic(Severity::Error, defaultValue,
                                 "Expected 'true' or 'false'"));
  }
  this->metadata->registerDiagnostic(
      defaultValue,
      Diagnostic(
          Severity::Warning, defaultValue,
          "String literals as value for boolean options are deprecated."));
}

void OptionDiagnosticVisitor::validateFeatureOption(Node *defaultValue) const {
  const auto *slNode = dynamic_cast<StringLiteral *>(defaultValue);
  if (!slNode) {
    this->metadata->registerDiagnostic(
        defaultValue,
        Diagnostic(Severity::Error, defaultValue, "Expected string"));
    return;
  }
  const auto &contents = slNode->id;
  if (contents == "enabled" || contents == "disabled" || contents == "auto") {
    return;
  }
  this->metadata->registerDiagnostic(
      defaultValue,
      Diagnostic(Severity::Error, defaultValue,
                 "Expected one of: 'enabled', 'disabled', 'auto'"));
}

void OptionDiagnosticVisitor::validateComboOption(Node *defaultValue,
                                                  ArgumentList *al) const {
  const auto *defaultSL = dynamic_cast<StringLiteral *>(defaultValue);
  std::optional<std::string> defaultString;
  if (!defaultSL) {
    this->metadata->registerDiagnostic(
        defaultValue,
        Diagnostic(Severity::Error, defaultValue, "Expected string literal"));
  } else {
    defaultString = defaultSL->id;
  }
  auto choicesKwarg = al->getKwarg("choices");
  if (!choicesKwarg.has_value()) {
    this->metadata->registerDiagnostic(
        al->parent,
        Diagnostic(Severity::Error, al->parent, "Missing 'choices' kwarg"));
    return;
  }
  std::set<std::string> foundChoices;
  const auto *choicesArray =
      dynamic_cast<ArrayLiteral *>(choicesKwarg.value().get());
  if (!choicesArray) {
    this->metadata->registerDiagnostic(
        choicesKwarg->get(), Diagnostic(Severity::Error, choicesKwarg->get(),
                                        "Expected array of strings"));
    return;
  }
  for (const auto &choice : choicesArray->args) {
    const auto *asLiteral = dynamic_cast<StringLiteral *>(choice.get());
    if (!asLiteral) {
      this->metadata->registerDiagnostic(
          choice.get(),
          Diagnostic(Severity::Error, choice.get(), "Expected string literal"));
      continue;
    }
    const auto &content = asLiteral->id;
    if (foundChoices.contains(content)) {
      this->metadata->registerDiagnostic(
          choice.get(), Diagnostic(Severity::Warning, choice.get(),
                                   "Duplicate choice '" + content + "'"));
      continue;
    }
    foundChoices.insert(content);
  }
  if (defaultString.has_value() &&
      !foundChoices.contains(defaultString.value())) {
    this->metadata->registerDiagnostic(
        defaultValue,
        Diagnostic(Severity::Error, defaultValue,
                   "Default value is not contained in the choices array."));
  }
}

void OptionDiagnosticVisitor::extractArrayChoices(
    ArgumentList *al, std::set<std::string> **choices) const {
  auto kwarg = al->getKwarg("choices");
  if (!kwarg.has_value()) {
    *choices = nullptr;
    return;
  }
  const auto *arrayLit = dynamic_cast<ArrayLiteral *>(kwarg->get());
  if (!arrayLit) {
    *choices = nullptr;
    this->metadata->registerDiagnostic(
        kwarg->get(),
        Diagnostic(Severity::Error, kwarg->get(), "Expected array literal"));
    return;
  }
  auto *ret = new std::set<std::string>();
  for (const auto &node : arrayLit->args) {
    const auto *asStr = dynamic_cast<StringLiteral *>(node.get());
    if (!asStr) {
      this->metadata->registerDiagnostic(
          node.get(),
          Diagnostic(Severity::Error, node.get(), "Expected string literal"));
      continue;
    }
    const auto &content = asStr->id;
    if (!ret->contains(content)) {
      ret->insert(content);
      continue;
    }
    this->metadata->registerDiagnostic(
        node.get(),
        Diagnostic(Severity::Warning, node.get(), "Duplicate choice"));
  }
  *choices = ret;
}

void OptionDiagnosticVisitor::validateArrayOption(Node *defaultValue,
                                                  ArgumentList *al) const {
  std::set<std::string> *choices = nullptr;
  extractArrayChoices(al, &choices);
  const auto *arrLit = dynamic_cast<ArrayLiteral *>(defaultValue);
  if (!arrLit) {
    this->metadata->registerDiagnostic(
        defaultValue,
        Diagnostic(Severity::Error, defaultValue, "Expected array literal"));
    goto end;
  }
  for (const auto &node : arrLit->args) {
    const auto *asStr = dynamic_cast<StringLiteral *>(node.get());
    if (!asStr) {
      this->metadata->registerDiagnostic(
          node.get(),
          Diagnostic(Severity::Error, node.get(), "Expected string literal"));
      continue;
    }
    const auto &content = asStr->id;
    if (!choices || choices->contains(content)) {
      continue;
    }
    this->metadata->registerDiagnostic(
        node.get(), Diagnostic(Severity::Error, node.get(),
                               "Value is not a valid choice!"));
  }
end:
  delete choices;
}

std::optional<int64_t>
OptionDiagnosticVisitor::fetchIntOrNullOpt(ArgumentList *al,
                                           const std::string &kwarg) {
  const auto &kwargNode = al->getKwarg(kwarg);
  if (!kwargNode.has_value()) {
    return std::nullopt;
  }
  const auto *node = kwargNode.value().get();
  auto val = this->parseInt(node);
  if (val.has_value()) {
    return val;
  }
  val = this->parseString(node);
  if (val.has_value()) {
    return val;
  }
  this->metadata->registerDiagnostic(
      node,
      Diagnostic(Severity::Error, node, "Unable to parse as integer literal"));
  return std::nullopt;
}

void OptionDiagnosticVisitor::validateIntegerOption(ArgumentList *al,
                                                    Node *defaultValue) {
  auto defaultInt = this->parseInt(defaultValue);
  if (!defaultInt.has_value()) {
    defaultInt = this->parseString(defaultValue);
    if (!defaultInt.has_value()) {
      this->metadata->registerDiagnostic(
          defaultValue, Diagnostic(Severity::Error, defaultValue,
                                   "Unable to parse as integer literal"));
    }
  }
  auto minValue = this->fetchIntOrNullOpt(al, "min");
  auto maxValue = this->fetchIntOrNullOpt(al, "max");

  if (!minValue.has_value() && !maxValue.has_value() &&
      !defaultInt.has_value()) {
    return;
  }
  if (minValue.has_value() && maxValue.has_value()) {
    auto minV = minValue.value();
    auto maxV = maxValue.value();
    const auto *minKwarg = al->getKwarg("min").value().get();
    if (minV > maxV) {
      this->metadata->registerDiagnostic(
          minKwarg,
          Diagnostic(Severity::Warning, minKwarg,
                     "Minimum value is greater than the maximum value"));
    } else if (minV == maxV) {
      this->metadata->registerDiagnostic(
          minKwarg, Diagnostic(Severity::Warning, minKwarg,
                               "Minimum value is equals to the maximum value"));
    }
  }
  if (minValue.has_value() && defaultInt.has_value()) {
    auto minV = minValue.value();
    auto defaultV = defaultInt.value();
    if (minV > defaultV) {
      this->metadata->registerDiagnostic(
          defaultValue,
          Diagnostic(Severity::Warning, defaultValue,
                     "Default value is lower than the minimum value"));
    }
  }
  if (maxValue.has_value() && defaultInt.has_value()) {
    auto maxV = maxValue.value();
    auto defaultV = defaultInt.value();
    if (maxV < defaultV) {
      this->metadata->registerDiagnostic(
          defaultValue,
          Diagnostic(Severity::Warning, defaultValue,
                     "Default value is greater than the maximum value"));
    }
  }
}

void OptionDiagnosticVisitor::visitFunctionExpression(
    FunctionExpression *node) {
  node->visitChildren(this);
  if (node->functionName() != "option") {
    this->metadata->registerDiagnostic(
        node, Diagnostic(Severity::Error, node,
                         "Invalid function call in meson options file: " +
                             node->functionName()));
    return;
  }
  auto *al = dynamic_cast<ArgumentList *>(node->args.get());
  if (!al) {
    this->metadata->registerDiagnostic(
        node, Diagnostic(Severity::Error, node,
                         "Missing arguments in call to `option`"));
    return;
  }
  auto nameNode = al->getPositionalArg(0);
  if (!nameNode.has_value()) {
    this->metadata->registerDiagnostic(
        node, Diagnostic(Severity::Error, node, "Missing option name"));
    return;
  }
  auto *nameNodeSl = dynamic_cast<StringLiteral *>(nameNode->get());
  if (!nameNodeSl) {
    this->metadata->registerDiagnostic(
        nameNode->get(), Diagnostic(Severity::Error, nameNode->get(),
                                    "Expected string literal"));
    return;
  }
  this->checkName(nameNodeSl);
  auto optionTypeNode = al->getKwarg("type");
  if (!optionTypeNode.has_value()) {
    this->metadata->registerDiagnostic(
        nameNode->get(), Diagnostic(Severity::Error, nameNode->get(),
                                    "Missing option type kwarg"));
    return;
  }
  auto *optionTypeSL = dynamic_cast<StringLiteral *>(optionTypeNode->get());
  if (!optionTypeSL) {
    this->metadata->registerDiagnostic(
        nameNode->get(),
        Diagnostic(Severity::Error, nameNode->get(),
                   "Expected option type to be a string literal"));
    return;
  }
  const auto &optionType = optionTypeSL->id;
  if (optionType != "string" && optionType != "integer" &&
      optionType != "boolean" && optionType != "combo" &&
      optionType != "array" && optionType != "feature") {
    this->metadata->registerDiagnostic(
        nameNode->get(), Diagnostic(Severity::Error, nameNode->get(),
                                    "Unknown option type: " + optionType));
  }
  auto defaultValue = al->getKwarg("value");
  if (!defaultValue.has_value()) {
    // TODO: Check e.g. min/max for integer
    return;
  }
  if (optionType == "string") {
    validateStringOption(defaultValue.value().get());
  } else if (optionType == "integer") {
    validateIntegerOption(al, defaultValue.value().get());
  } else if (optionType == "boolean") {
    validateBooleanOption(defaultValue.value().get());
  } else if (optionType == "feature") {
    validateFeatureOption(defaultValue.value().get());
  } else if (optionType == "combo") {
    validateComboOption(defaultValue.value().get(), al);
  } else if (optionType == "array") {
    validateArrayOption(defaultValue.value().get(), al);
  }
}

void OptionDiagnosticVisitor::visitIdExpression(IdExpression *node) {
  node->visitChildren(this);
}

void OptionDiagnosticVisitor::visitIntegerLiteral(IntegerLiteral *node) {
  node->visitChildren(this);
}

void OptionDiagnosticVisitor::visitIterationStatement(
    IterationStatement *node) {
  node->visitChildren(this);
}

void OptionDiagnosticVisitor::visitKeyValueItem(KeyValueItem *node) {
  node->visitChildren(this);
}

void OptionDiagnosticVisitor::visitKeywordItem(KeywordItem *node) {
  node->visitChildren(this);
}

void OptionDiagnosticVisitor::visitMethodExpression(MethodExpression *node) {
  node->visitChildren(this);
}

void OptionDiagnosticVisitor::visitSelectionStatement(
    SelectionStatement *node) {
  node->visitChildren(this);
}

void OptionDiagnosticVisitor::visitStringLiteral(StringLiteral *node) {
  node->visitChildren(this);
}

void OptionDiagnosticVisitor::visitSubscriptExpression(
    SubscriptExpression *node) {
  node->visitChildren(this);
}

void OptionDiagnosticVisitor::visitUnaryExpression(UnaryExpression *node) {
  node->visitChildren(this);
}

void OptionDiagnosticVisitor::visitErrorNode(ErrorNode *node) {
  node->visitChildren(this);
  this->metadata->registerDiagnostic(
      node, Diagnostic(Severity::Error, node, node->message));
}

void OptionDiagnosticVisitor::visitBreakNode(BreakNode *node) {
  node->visitChildren(this);
}

void OptionDiagnosticVisitor::visitContinueNode(ContinueNode *node) {
  node->visitChildren(this);
}
